server: TestServerControllerWaitForDefaultCluster failed

server.TestServerControllerWaitForDefaultCluster [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12548632?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12548632?buildTab=artifacts#/) on master @ [7a0b7f4e4929f8fb3e91a6b82290a0857f33b41e](https://github.com/cockroachdb/cockroach/commits/7a0b7f4e4929f8fb3e91a6b82290a0857f33b41e):


```
=== RUN   TestServerControllerWaitForDefaultCluster
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster2924574746
    test_log_scope.go:81: use -show-logs to present logs inline
    server_controller_test.go:158: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server/server_controller_test.go:158
        	Error:      	Received unexpected error:
        	            	pq: service unavailable for target tenant (hello)
        	Test:       	TestServerControllerWaitForDefaultCluster
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster2924574746
--- FAIL: TestServerControllerWaitForDefaultCluster (19.55s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/multi-tenant @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerWaitForDefaultCluster.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33270

server.TestServerControllerWaitForDefaultCluster [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12568690?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12568690?buildTab=artifacts#/) on master @ [013c7e6703244f541bece6a1037ccad080eb7616](https://github.com/cockroachdb/cockroach/commits/013c7e6703244f541bece6a1037ccad080eb7616):


```
=== RUN   TestServerControllerWaitForDefaultCluster
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster313079808
    test_log_scope.go:81: use -show-logs to present logs inline
    server_controller_test.go:158: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server/server_controller_test.go:158
        	Error:      	Received unexpected error:
        	            	pq: service unavailable for target tenant (hello)
        	Test:       	TestServerControllerWaitForDefaultCluster
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster313079808
--- FAIL: TestServerControllerWaitForDefaultCluster (17.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerWaitForDefaultCluster.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestServerControllerWaitForDefaultCluster [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12590149?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12590149?buildTab=artifacts#/) on master @ [e43bfbd92f006876719f1600b4ac3b094b85835f](https://github.com/cockroachdb/cockroach/commits/e43bfbd92f006876719f1600b4ac3b094b85835f):


```
=== RUN   TestServerControllerWaitForDefaultCluster
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster1511968327
    test_log_scope.go:81: use -show-logs to present logs inline
    server_controller_test.go:158: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server/server_controller_test.go:158
        	Error:      	Received unexpected error:
        	            	pq: service unavailable for target tenant (hello)
        	Test:       	TestServerControllerWaitForDefaultCluster
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster1511968327
--- FAIL: TestServerControllerWaitForDefaultCluster (19.80s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #114171 server: TestServerControllerWaitForDefaultCluster failed [C-test-failure O-robot T-multitenant branch-release-23.2 release-blocker]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerWaitForDefaultCluster.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

server.TestServerControllerWaitForDefaultCluster [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12609834?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12609834?buildTab=artifacts#/) on master @ [11fb3f2e98ff6ad59ad67f05152c9d16525bb77b](https://github.com/cockroachdb/cockroach/commits/11fb3f2e98ff6ad59ad67f05152c9d16525bb77b):


```
=== RUN   TestServerControllerWaitForDefaultCluster
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster2828121859
    test_log_scope.go:81: use -show-logs to present logs inline
    server_controller_test.go:158: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/server/server_controller_test.go:158
        	Error:      	Received unexpected error:
        	            	pq: service unavailable for target tenant (hello)
        	Test:       	TestServerControllerWaitForDefaultCluster
    panic.go:523: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/5b2c9b3a394428c7572d34050aad8975/logTestServerControllerWaitForDefaultCluster2828121859
--- FAIL: TestServerControllerWaitForDefaultCluster (19.71s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #114171 server: TestServerControllerWaitForDefaultCluster failed [C-test-failure O-robot T-multitenant branch-release-23.2]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerControllerWaitForDefaultCluster.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/server/server_controller_test.go
+++ b/pkg/server/server_controller_test.go
@@ -21,8 +21,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
-	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
@@ -139,7 +139,7 @@ func TestServerControllerWaitForDefaultCluster(t *testing.T) {
 	defer s.Stopper().Stop(ctx)
 
 	sqlRunner := sqlutils.MakeSQLRunner(db)
-	if util.RaceEnabled {
+	if skip.Duress() {
 		sqlRunner.Exec(t, "SET CLUSTER SETTING server.controller.mux_virtual_cluster_wait.timeout = '1m'")
 	}
 

--- a/pkg/testutils/skip/skip.go
+++ b/pkg/testutils/skip/skip.go
@@ -32,9 +32,8 @@ type SkippableTest interface {
 // WithIssue skips this test, logging the given issue ID as the reason.
 func WithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
-	maybeSkip(t, append([]interface{}{
-		fmt.Sprintf("https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID)},
-		args...))
+	maybeSkip(t, fmt.Sprintf("https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID),
+		args...)
 }
 
 // Unimplemented skips this test case, loggint the given issue ID. It
@@ -43,20 +42,22 @@ func WithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 // because of an unimplemented feature.
 func Unimplemented(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
-	maybeSkip(t, append([]interface{}{
-		fmt.Sprintf("https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID)},
-		args...))
+	maybeSkip(t, withIssue("unimplemented", githubIssueID), args...)
 }
 
 // IgnoreLint skips this test, explicitly marking it as not a test that
 // should be tracked as a "skipped test" by external tools. You should use this
 // if, for example, your test should only be run in Race mode.
+//
+// Does not respect COCKROACH_FORCE_RUN_SKIPPED_TESTS.
 func IgnoreLint(t SkippableTest, args ...interface{}) {
 	t.Helper()
-	maybeSkip(t, args...)
+	t.Skip(args...)
 }
 
 // IgnoreLintf is like IgnoreLint, and it also takes a format string.
+//
+// Does not respect COCKROACH_FORCE_RUN_SKIPPED_TESTS.
 func IgnoreLintf(t SkippableTest, format string, args ...interface{}) {
 	t.Helper()
 	t.Skipf(format, args...)
@@ -66,7 +67,7 @@ func IgnoreLintf(t SkippableTest, format string, args ...interface{}) {
 func UnderDeadlock(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if syncutil.DeadlockEnabled {
-		maybeSkip(t, append([]interface{}{"disabled under deadlock detector"}, args...))
+		maybeSkip(t, "disabled under deadlock detector", args...)
 	}
 }
 
@@ -75,10 +76,7 @@ func UnderDeadlock(t SkippableTest, args ...interface{}) {
 func UnderDeadlockWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
 	if syncutil.DeadlockEnabled {
-		maybeSkip(t, append([]interface{}{fmt.Sprintf(
-			"disabled under deadlock detector. issue: https://github.com/cockroachdb/cockroach/issues/%d",
-			githubIssueID,
-		)}, args...))
+		maybeSkip(t, withIssue("disabled under deadlock detector", githubIssueID), args...)
 	}
 }
 
@@ -86,7 +84,7 @@ func UnderDeadlockWithIssue(t SkippableTest, githubIssueID int, args ...interfac
 func UnderRace(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if util.RaceEnabled {
-		maybeSkip(t, append([]interface{}{"disabled under race"}, args...))
+		maybeSkip(t, "disabled under race", args...)
 	}
 }
 
@@ -95,9 +93,7 @@ func UnderRace(t SkippableTest, args ...interface{}) {
 func UnderRaceWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
 	if util.RaceEnabled {
-		maybeSkip(t, append([]interface{}{fmt.Sprintf(
-			"disabled under race. issue: https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID,
-		)}, args...))
+		maybeSkip(t, withIssue("disabled under race", githubIssueID), args...)
 	}
 }
 
@@ -106,9 +102,7 @@ func UnderRaceWithIssue(t SkippableTest, githubIssueID int, args ...interface{})
 func UnderBazelWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
 	if bazel.BuiltWithBazel() {
-		maybeSkip(t, append([]interface{}{fmt.Sprintf(
-			"disabled under bazel. issue: https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID,
-		)}, args...))
+		maybeSkip(t, withIssue("disabled under bazel", githubIssueID), args...)
 	}
 }
 
@@ -119,7 +113,7 @@ var _ = UnderBazelWithIssue
 func UnderShort(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if testing.Short() {
-		maybeSkip(t, append([]interface{}{"disabled under -short"}, args...))
+		maybeSkip(t, "disabled under -short", args...)
 	}
 }
 
@@ -127,7 +121,7 @@ func UnderShort(t SkippableTest, args ...interface{}) {
 func UnderStress(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if Stress() {
-		maybeSkip(t, append([]interface{}{"disabled under stress"}, args...))
+		maybeSkip(t, "disabled under stress", args...)
 	}
 }
 
@@ -136,9 +130,7 @@ func UnderStress(t SkippableTest, args ...interface{}) {
 func UnderStressWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
 	if Stress() {
-		maybeSkip(t, append([]interface{}{fmt.Sprintf(
-			"disabled under stress. issue: https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID,
-		)}, args...))
+		maybeSkip(t, withIssue("disabled under stress", githubIssueID), args...)
 	}
 }
 
@@ -147,7 +139,7 @@ func UnderStressWithIssue(t SkippableTest, githubIssueID int, args ...interface{
 func UnderStressRace(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if Stress() && util.RaceEnabled {
-		maybeSkip(t, append([]interface{}{"disabled under stressrace"}, args...))
+		maybeSkip(t, "disabled under stressrace", args...)
 	}
 }
 
@@ -156,7 +148,7 @@ func UnderStressRace(t SkippableTest, args ...interface{}) {
 func UnderMetamorphic(t SkippableTest, args ...interface{}) {
 	t.Helper()
 	if util.IsMetamorphicBuild() {
-		maybeSkip(t, append([]interface{}{"disabled under metamorphic"}, args...))
+		maybeSkip(t, "disabled under metamorphic", args...)
 	}
 }
 
@@ -166,9 +158,7 @@ func UnderMetamorphic(t SkippableTest, args ...interface{}) {
 func UnderMetamorphicWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
 	t.Helper()
 	if util.IsMetamorphicBuild() {
-		maybeSkip(t, append([]interface{}{fmt.Sprintf(
-			"disabled under metamorphic. issue: https://github.com/cockroachdb/cockroach/issues/%d", githubIssueID,
-		)}, args...))
+		maybeSkip(t, withIssue("disabled under metamorphic", githubIssueID), args...)
 	}
 }
 
@@ -191,12 +181,20 @@ func UnderBench() bool {
 	return f != nil && f.Value.String() != ""
 }
 
+func withIssue(reason string, githubIssueID int) string {
+	return fmt.Sprintf(
+		"%s. issue: https://github.com/cockroachdb/cockroach/issues/%d",
+		reason,
+		githubIssueID,
+	)
+}
+
 var forceRunSkippedTests = envutil.EnvOrDefaultBool("COCKROACH_FORCE_RUN_SKIPPED_TESTS", false)
 
-func maybeSkip(t SkippableTest, args ...interface{}) {
+func maybeSkip(t SkippableTest, reason string, args ...interface{}) {
 	if forceRunSkippedTests {
 		return
 	}
 
-	t.Skip(args...)
+	t.Skip(append([]interface{}{reason}, args...)...)
 }

--- a/pkg/testutils/skip/skip.go
+++ b/pkg/testutils/skip/skip.go
@@ -13,6 +13,7 @@ package skip
 import (
 	"flag"
 	"fmt"
+	"strings"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/build/bazel"
@@ -170,6 +171,22 @@ func UnderNonTestBuild(t SkippableTest) {
 	}
 }
 
+// UnderDuress skips the test if we are running under any of the
+// conditions we have observed as producing slow builds.
+func UnderDuressWithIssue(t SkippableTest, githubIssueID int, args ...interface{}) {
+	t.Helper()
+	if Duress() {
+		skipReason := fmt.Sprintf("duress (current config %s)", testConfig())
+		maybeSkip(t, withIssue(skipReason, githubIssueID), args...)
+	}
+}
+
+// Duress catures the conditions that currently lead us to
+// believe that tests may be slower than normal.
+func Duress() bool {
+	return util.RaceEnabled || Stress() || syncutil.DeadlockEnabled
+}
+
 // UnderBench returns true iff a test is currently running under `go
 // test -bench`.  When true, tests should avoid writing data on
 // stdout/stderr from goroutines that run asynchronously with the
@@ -179,6 +196,20 @@ func UnderBench() bool {
 	// test executable with `-test.bench 1`.
 	f := flag.Lookup("test.bench")
 	return f != nil && f.Value.String() != ""
+}
+
+func testConfig() string {
+	configs := []string{}
+	if Stress() {
+		configs = append(configs, "stress")
+	}
+	if util.RaceEnabled {
+		configs = append(configs, "race")
+	}
+	if syncutil.DeadlockEnabled {
+		configs = append(configs, "deadlock")
+	}
+	return strings.Join(configs, ",")
 }
 
 func withIssue(reason string, githubIssueID int) string {

